### PR TITLE
[A2-863] add db_id to tokens table

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

To start, this migrates our hand-rolled migration code to using the library we use in other places (see also https://github.com/chef/automate/pull/665).

#### TODO

- [X] introduce `db_id` into the tokens table
- [ ] figure out if there's indices beyond `db_id` that we would care about
  - [ ] `value`, maybe? we'll want this to **lookup quickly** (removed this for the moment, but if we bring it back to this PR, docs suggest it will [automatically be indexed](https://www.postgresql.org/docs/9.6/indexes-unique.html) -- blake. ... (actually, regardless of the uniqueness, I think this is worth indexing, but it might make sense to wait for adding uniqueness since I expect it could be slow)
  - [x] also `id` uniqueness, given the tests: [here](https://buildkite.com/chef/chef-automate-master-verify-private/builds/2420#8e0f2f75-d63a-42c9-82e1-a443196f6d46/681-791) (confused by this since `primary_key` should have created uniqueness automatically on `id` 🤔-blake )

### :+1: Definition of Done

After deploying, the tokens table will have a `db_id` column.

### :athletic_shoe: Demo Script / Repro Steps

- `start_all_services`
- create a token: `chef-automate iam token create foobar`
- check the database:
```
chef_authn_service^# select * from chef_authn_tokens
   id   |            value             |            created            |            updated            | active | description | project_ids
--------+------------------------------+-------------------------------+-------------------------------+--------+-------------+-------------
 foobar | PgMxj3m1WNH7XukboIJRaHCyT0k= | 2019-06-24 15:24:38.565056+00 | 2019-06-24 15:24:38.565056+00 | t      | foobar      | {}
(1 row)

chef_authn_service^# select * from migrations ;
 num |                      descr                       |              at
-----+--------------------------------------------------+-------------------------------
   1 | create clients table                             | 2019-06-24 15:23:15.020386+00
   2 | add 'active' column to clients table             | 2019-06-24 15:23:15.037881+00
   3 | change id to type UUID                           | 2019-06-24 15:23:15.072423+00
   4 | add 'description' column to clients table        | 2019-06-24 15:23:15.108413+00
   5 | rename clients table tokens table                | 2019-06-24 15:23:15.131061+00
   6 | rename 'token' column of tokens table to 'value' | 2019-06-24 15:23:15.141391+00
   7 | change id to type text                           | 2019-06-24 15:23:15.151059+00
   8 | add column project_ids                           | 2019-06-24 15:23:15.196293+00
   9 | remove default project_ids                       | 2019-06-24 15:23:15.276301+00
  10 | drop constraint                                  | 2019-06-24 15:23:15.288506+00
  11 | add default empty array for project_ids          | 2019-06-24 15:23:15.294687+00
  12 | add filter function for token projects           | 2019-06-24 15:23:15.306849+00
(12 rows)

chef_authn_service^#
```
- `rebuild components/authn-service`
- check the database:
```
chef_authn_service^# select * from migrations ;
ERROR:  relation "migrations" does not exist
LINE 1: select * from migrations ;
                      ^
chef_authn_service^# select * from chef_authn_tokens;
   id   |            value             |            created            |            updated            | active | description | project_ids | db_id
--------+------------------------------+-------------------------------+-------------------------------+--------+-------------+-------------+-------
 foobar | PgMxj3m1WNH7XukboIJRaHCyT0k= | 2019-06-24 15:24:38.565056+00 | 2019-06-24 15:24:38.565056+00 | t      | foobar      | {}          |     1
(1 row)

chef_authn_service^# select * from schema_migrations ;
 version | dirty
---------+-------
      13 | f
(1 row)

chef_authn_service^#
```

### :chains: Related Resources

- [A2-863](https://chefio.atlassian.net/browse/A2-863)

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [ ] Code actually executed?
- [ ] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authn-service/tokens/pg/migrate.go [hunk 1]
```diff
@@ -1,183 +1,76 @@
...
+	// When this is done, we drop that table, so the code here will only run once.
+	// We then run the usual "up" migration code, which will pick up where we've left off.
+	// Existing migrations have been converted into up.sql files accordingly, keeping
+	// the numbering intact.
+
+	ctx := context.TODO()
+	tx, err := db.BeginTx(ctx, nil)
 	if err != nil {
-		return errors.Wrapf(err, "acquiring advisory lock %d", constants.AdvisoryLockID)
+		return errors.Wrap(err, "being transaction")
```
Comment (blakestier) [inline components/authn-service/tokens/pg/migrate.go]: typo here: being -> beginning

File: components/authn-service/tokens/pg/sql/02_add__active__column_to_clients_table.up.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE chef_authn_clients ADD COLUMN active BOOLEAN NOT NULL DEFAULT TRUE;
```
Comment (blakestier) [inline components/authn-service/tokens/pg/sql/02_add__active__column_to_clients_table.up.sql]: 🤔 why isn't github highlighting this like the others?

File: components/authn-service/tokens/pg/sql/12_add_filter_function_for_token_projects.up.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (blakestier) [inline components/authn-service/tokens/pg/sql/12_add_filter_function_for_token_projects.up.sql]: extra line? also, does that even matter?

Comment (srenatus) [reply components/authn-service/tokens/pg/sql/12_add_filter_function_for_token_projects.up.sql]: I've mechanically translated the inline statements in our go code into these files -- so, that one was probably like
```go
Exec(`
SOME SQL HERE
`
)
```
and that backtick right behind the `(` was what gave us this newline. 🤷‍♂ I think it doesn't matter 😃

File: components/authn-service/tokens/pg/sql/02_add__active__column_to_clients_table.up.sql (hunk 2 above)
Comment (srenatus) [reply components/authn-service/tokens/pg/sql/02_add__active__column_to_clients_table.up.sql]: I have no idea 😅

File: components/authn-service/tokens/pg/migrate.go (hunk 1 above)
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: I've wanted to say "begin", but "beginning" works for me, too. 😃

**Comment by blakestier** [comment]
some tests breaking (reasonably). I'll try to get this green today

File: components/authn-service/tokens/pg/migrate.go [hunk 4]
```diff
@@ -1,183 +1,76 @@
...
 
-		if done {
-			break
-		}
-		i++ // rinse and repeat for next migration
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "commit migration table migration")
 	}
 
+	m, err := migrate.New("file://"+path, url)
```
Comment (msorens) [inline components/authn-service/tokens/pg/migrate.go]: This is done in 6 other places, all of which define their own copy of `addScheme` and use that, so might be time to add `addScheme` as a library helper function...

File: components/authn-service/tokens/pg/migrate.go [hunk 5]
```diff
@@ -1,183 +1,76 @@
 package pg
 
 import (
+	"context"
 	"database/sql"
-	"fmt"
 
+	"github.com/mattes/migrate"
+	_ "github.com/mattes/migrate/database/postgres" // make database available
+	_ "github.com/mattes/migrate/source/file"       // make source available
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: Yeah but it's not been merged, I believe. Either one will need a fix in rebase, whichever merges last.

File: components/authn-service/tokens/pg/migrate.go [hunk 6]
```diff
@@ -1,183 +1,76 @@
...
-	{
-		descr:     "rename 'token' column of tokens table to 'value'",
-		statement: `ALTER TABLE chef_authn_tokens RENAME COLUMN token TO value;`,
-	},
-	{
-		descr:     "change id to type text",
-		statement: `ALTER TABLE chef_authn_tokens ALTER COLUMN id TYPE TEXT USING id::TEXT;`,
-	},
-	// This migration will add '{"default"}' project in everywhere, as old tokens need
-	// to start with a project.
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: Good point. Now that's manual work ;)

File: components/authn-service/tokens/pg/migrate.go [hunk 7]
```diff
@@ -1,183 +1,76 @@
...
 
-// migrate tries to execute all the migrations we know of
-func migrate(c *sql.DB) error {
-	// Acquire advisory lock
-	_, err := c.Exec("SELECT pg_advisory_lock($1);", constants.AdvisoryLockID)
+// runMigrations tries to execute all the migrations we know of
+func runMigrations(db *sql.DB, url, path string) error {
+	// We're migrating to golang-migrate/migrate instead of this hand-rolled code:
+	// If it exists, we're migrating the existing migration table a golang-migrate conform
+	// schema_migrations table.
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: Missing a "to" indeed

File: components/authn-service/habitat/plan.sh [hunk 8]
```diff
@@ -52,3 +52,11 @@ do_prepare() {
...
     build_line "Setting GO_LDFLAGS=${GO_LDFLAGS}"
 }
+
+do_install() {
+  # Scaffolding go install callback
+  scaffolding_go_install
+
+  build_line "Copying migration files"
+  cp -r tokens/pg/sql "${pkg_prefix}/migrations"
+}
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/habitat/plan.sh]: Putting the migration sql files into the hart, so they can be put into the right place when running the service.

File: components/authn-service/tokens/pg/migrate.go (hunk 4 above)
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: ....or just do the simple thing in the other places. Having looked at the `addScheme` helper after some time, it seems a bit over-engineered. 😉

**Comment by srenatus** [comment]
> confused by this since `primary_key` should have created uniqueness automatically on `id` 🤔-blake 

@blakestier it did _before_ -- now we're using `db_id` as primary key, so `id` no longer is our PK. To keep uniqueness, we'll need the extra constraint on `id`. (Contrary to `value`, `id` cannot have duplicates right now, so what we do is safe.)

File: components/authn-service/tokens/pg/migrate.go [hunk 9]
```diff
@@ -1,183 +1,76 @@
...
+  IF EXISTS(SELECT * FROM information_schema.tables
+              WHERE table_schema = current_schema()
+              AND table_name = 'migrations') THEN
+     INSERT INTO schema_migrations (
+       SELECT MAX(num) AS version, FALSE AS dirty FROM migrations
+     );
+  END IF;
+END;
+$$
+LANGUAGE plpgsql;`); err != nil {
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: `define_move_if_exists` would violate naming conventions, but I've added a constant 👍

File: components/authn-service/tokens/pg/migrate.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: ✅ the other PR is merged, this is adapted.

File: components/authn-service/tokens/pg/migrate.go (hunk 4 above)
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: I've decided to do a cleanup sweep instead: https://github.com/chef/automate/compare/2ac97586b5bb93f26ab65acf78990e6c12479896..sr/a2-863/add-db_id-to-tokens-table

File: scripts/repo_health [hunk 10]
```diff
@@ -35,7 +35,7 @@ done
 echo "Checking for up-to-date bldr config"
 go run ./tools/bldr-config-gen
 if ! git diff --exit-code --ignore-submodules=all; then
-    echo "The blr config appears to be out of date!"
+    echo "The bldr config appears to be out of date!"
```
Comment (stevendanna) [inline scripts/repo_health]: :D Thanks

File: components/authn-service/habitat/hooks/run [hunk 11]
```diff
@@ -23,5 +23,7 @@ addNoProxy {{teams-service.sys.ip}}
 {{~/if}}
 {{~/eachAlive}}
 
+rm -rf {{pkg.svc_static_path}}/migrations
+cp -r {{pkg.path}}/migrations {{pkg.svc_static_path}}
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/habitat/hooks/run]: Hah, no idea. Having a hard time guessing what's cargo-culted and what's necessary here 😅

File: components/authn-service/tokens/pg/migrate.go [hunk 12]
```diff
@@ -1,183 +1,68 @@
...
-
-type migration struct {
-	descr     string
-	statement string
-}
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "commit migration table migration")
+	}
 
-var migrations = []migration{
```
[the remark this answers is not in the document]
Comment (srenatus) [reply components/authn-service/tokens/pg/migrate.go]: Thanks for the pointer -- I've opted for making this look the same among the auth services instead... now that it's all in one package, we might be having an easier time going forward 🤔

File: components/authz-service/storage/postgres/migration/migration.go [hunk 13]
```diff
@@ -19,48 +15,7 @@ type Config struct {
...
-	m, err := migrate.New(addScheme(c.Path), c.PGURL.String())
-	if err != nil {
-		return errors.Wrap(err, "init migrate")
-	}
-
-	m.Log = migrationLog{c.Logger} // nolint: govet
-
-	err = m.Up()
-	// Note: we've screwed up migration 25. If the state is dirty, and we're at
-	// version 25, reset to 24 and try again.
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/migration.go]: Wondering about this custom patch... I infer that, once we have created migrations 26 and above, this redo code is no longer needed...?

**Pull request merged**